File: Data/generatePlots.m
clear;
close all;
clc;

% MATLAB replacement for:
%   plot.gp
%   plotbranch.txt
%   plotK.txt
%   plotL.txt
%   plotTK.txt
%   plotTL.txt
%
% Run this script from the directory containing the .dat files.
%
% Outputs:
%   powerlawK.pdf
%   powerlawL.pdf
%   powerlawTK.pdf
%   powerlawTL.pdf
%   branch.pdf

%% Common parameters

ells = [1.05 1.20 1.35 1.50 1.65 1.80 1.95];

K_values = [1.0e4 3.2e4 1.0e5 3.2e5 1.0e6];
K_file_strings = { ...
    '1.0e+04', ...
    '3.2e+04', ...
    '1.0e+05', ...
    '3.2e+05', ...
    '1.0e+06'};

K_labels = { ...
    '10^{4.0}', ...
    '10^{4.5}', ...
    '10^{5.0}', ...
    '10^{5.5}', ...
    '10^{6.0}'};

fontSize = 20;
figPosition = [100 100 1200 800];

% MATLAB default color order gives a consistent distinct color per curve.
colors = lines(max(numel(ells), numel(K_values)));


%% ========================================================================
%  1. ENERGY SCALING WITH INITIAL ENERGY K
%     gnuplot: plotK.txt
%
%     Fit:
%         log10(K*) = b + a log10(K)
%     so:
%         K* ~ K^a
% ========================================================================

figure('Position', figPosition);
hold on;

aK = zeros(size(ells));
bK = zeros(size(ells));

for i = 1:numel(ells)

    ell = ells(i);

    filename = sprintf( ...
        'powerlawK_IC_s1_ell1_%.2f_ell2_%.2f.dat', ...
        ell, ell);

    data = readmatrix(filename);

    K = data(:,1);
    Kstar = data(:,7);

    % Remove invalid entries before taking logarithms.
    valid = isfinite(K) & isfinite(Kstar) & K > 0 & Kstar > 0;
    Kfit = K(valid);
    KstarFit = Kstar(valid);

    % Equivalent to:
    % fit f(x) ... using (log10($1)):(log10($7))
    p = polyfit(log10(Kfit), log10(KstarFit), 1);

    aK(i) = p(1);
    bK(i) = p(2);

    loglog( ...
        K, Kstar, '-o', ...
        'Color', colors(i,:), ...
        'LineWidth', 1.5, ...
        'MarkerSize', 6, ...
        'DisplayName', sprintf( ...
            '$\\widetilde K_{%.2f} \\approx 10^{%.3f} K^{%.3f}$ \\,', ...
            ell, bK(i), aK(i)));
end

powerlawK = mean(aK);
stdevK = std(aK, 0);   % sample standard deviation, denominator N-1

xlabel('Initial Energy $K$', ...
    'Interpreter', 'latex', ...
    'FontSize', fontSize);

ylabel('Peak Transient Energy $\widetilde{K}_{\ell}$', ...
    'Interpreter', 'latex', ...
    'FontSize', fontSize);

title('$L^2$ Energy-Scaling Law for 2D Kuramoto-Sivashinsky', ...
    'Interpreter', 'latex', ...
    'FontSize', fontSize);

subtitle(sprintf([ ...
    '$K=[10^4,10^{4.5},\\ldots,10^6]$ and ' ...
    '$\\ell=[1.05,1.20,\\ldots,1.95]$: ' ...
    '$\\widetilde{K}_{\\ell} \\sim K^{%.4f \\pm %.4f}$'], ...
    powerlawK, stdevK), ...
    'Interpreter', 'latex', ...
    'FontSize', fontSize);

grid on;
box on;
xlim([1e4 1e6]);

set(gca, ...
    'XScale', 'log', ...
    'YScale', 'log', ...
    'FontSize', fontSize);

legend( ...
    'Location', 'southeast', ...
    'FontSize', fontSize-4, ...
    'Interpreter', 'latex');

exportgraphics(gcf, 'powerlawK.pdf', 'ContentType', 'vector');

fprintf('Energy scaling K exponents:\n');
disp(aK);
fprintf('Mean exponent = %.8f\n', powerlawK);
fprintf('Sample std    = %.8f\n\n', stdevK);


%% ========================================================================
%  2. ENERGY SCALING WITH DOMAIN FACTOR ell
%     gnuplot: plotL.txt
%
%     Fit:
%         log10(K*) = b + a ell
%     so:
%         K* ~ 10^(a ell + b)
% ========================================================================

figure('Position', figPosition);
hold on;

aL = zeros(size(K_values));
bL = zeros(size(K_values));

for i = 1:numel(K_values)

    filename = sprintf( ...
        'powerlawL_IC_s1_K_%s.dat', ...
        K_file_strings{i});

    data = readmatrix(filename);

    ell = data(:,2);
    Kstar = data(:,7);

    valid = isfinite(ell) & isfinite(Kstar) & ell > 0 & Kstar > 0;
    ellFit = ell(valid);
    KstarFit = Kstar(valid);

    % Equivalent to:
    % fit f(x) ... using 2:(log10($7))
    p = polyfit(ellFit, log10(KstarFit), 1);

    aL(i) = p(1);
    bL(i) = p(2);

    semilogy( ...
        ell, Kstar, '-o', ...
        'Color', colors(i,:), ...
        'LineWidth', 1.5, ...
        'MarkerSize', 6, ...
        'DisplayName', sprintf( ...
            '$\\widetilde K_{%s} \\approx 10^{%.3f}10^{%.3f\\ell}$', ...
            K_labels{i}, bL(i), aL(i)));
end

powerlawL = mean(aL);
stdevL = std(aL, 0);

xlabel('Domain Factor $\ell$', ...
    'Interpreter', 'latex', ...
    'FontSize', fontSize);

ylabel('Peak Transient Energy $\widetilde{K}_{K}$', ...
    'Interpreter', 'latex', ...
    'FontSize', fontSize);

title('Isotropic Domain-Scaling Law for 2D Kuramoto-Sivashinsky', ...
    'Interpreter', 'latex', ...
    'FontSize', fontSize);

subtitle(sprintf([ ...
    '$K=[10^4,10^{4.5},\\ldots,10^6]$ and ' ...
    '$\\ell=[1.05,1.20,\\ldots,1.95]$: ' ...
    '$\\widetilde{K}_{K} \\sim 10^{%.4f\\ell \\pm %.4f}$'], ...
    powerlawL, stdevL), ...
    'Interpreter', 'latex', ...
    'FontSize', fontSize);

grid on;
box off;
xlim([1 2]);

set(gca, ...
    'XScale', 'linear', ...
    'YScale', 'log', ...
    'FontSize', fontSize);

legend( ...
    'Location', 'southoutside', ...
    'FontSize', fontSize-4, ...
    'Interpreter', 'latex',...
    'NumColumns', 3);

exportgraphics(gcf, 'powerlawL.pdf', 'ContentType', 'vector');

fprintf('Domain scaling energy exponents:\n');
disp(aL);
fprintf('Mean exponent = %.8f\n', powerlawL);
fprintf('Sample std    = %.8f\n\n', stdevL);


%% ========================================================================
%  3. TIME SCALING WITH INITIAL ENERGY K
%     gnuplot: plotTK.txt
%
%     Fit:
%         log10(T*) = b + a log10(K)
%     so:
%         T* ~ K^a
% ========================================================================

figure('Position', figPosition);
hold on;

aTK = zeros(size(ells));
bTK = zeros(size(ells));

for i = 1:numel(ells)

    ell = ells(i);

    filename = sprintf( ...
        'powerlawTK_IC_s1_ell1_%.2f_ell2_%.2f.dat', ...
        ell, ell);

    data = readmatrix(filename);

    K = data(:,1);
    Tstar = data(:,4);

    valid = isfinite(K) & isfinite(Tstar) & K > 0 & Tstar > 0;
    Kfit = K(valid);
    TstarFit = Tstar(valid);

    % Equivalent to:
    % fit f(x) ... using (log10($1)):(log10($4))
    p = polyfit(log10(Kfit), log10(TstarFit), 1);

    aTK(i) = p(1);
    bTK(i) = p(2);

    loglog( ...
        K, Tstar, '-o', ...
        'Color', colors(i,:), ...
        'LineWidth', 1.5, ...
        'MarkerSize', 6, ...
        'DisplayName', sprintf( ...
            '$\\widetilde{T}_{%.2f} \\approx 10^{%.3f}K^{%.3f}$', ...
            ell, bTK(i), aTK(i)));
end

powerlawTK = mean(aTK);
stdevTK = std(aTK, 0);

xlabel('Initial Energy $K$', ...
    'Interpreter', 'latex', ...
    'FontSize', fontSize);

ylabel('Time Window of Peak Transient Energy $\widetilde{T}_{\ell}$', ...
    'Interpreter', 'latex', ...
    'FontSize', fontSize);

title('$L^2$ Energy Time-Scaling Law for 2D Kuramoto-Sivashinsky', ...
    'Interpreter', 'latex', ...
    'FontSize', fontSize);

subtitle(sprintf([ ...
    '$K=[10^4,10^{4.5},\\ldots,10^6]$ and ' ...
    '$\\ell=[1.05,1.20,\\ldots,1.95]$: ' ...
    '$\\widetilde{T}_{\\ell} \\sim K^{%.4f \\pm %.4f}$'], ...
    powerlawTK, stdevTK), ...
    'Interpreter', 'latex', ...
    'FontSize', fontSize);

grid on;
box on;
xlim([1e4 1e6]);

set(gca, ...
    'XScale', 'log', ...
    'YScale', 'log', ...
    'FontSize', fontSize);

legend( ...
    'Location', 'southwest', ...
    'FontSize', fontSize-4, ...
    'Interpreter', 'latex');

exportgraphics(gcf, 'powerlawTK.pdf', 'ContentType', 'vector');

fprintf('Time-vs-K exponents:\n');
disp(aTK);
fprintf('Mean exponent = %.8f\n', powerlawTK);
fprintf('Sample std    = %.8f\n\n', stdevTK);


%% ========================================================================
%  4. TIME SCALING WITH DOMAIN FACTOR ell
%     gnuplot: plotTL.txt
%
%     Fit:
%         log10(T*) = b + a ell
%     so:
%         T* ~ 10^(a ell + b)
% ========================================================================

figure('Position', figPosition);
hold on;

aTL = zeros(size(K_values));
bTL = zeros(size(K_values));

for i = 1:numel(K_values)

    filename = sprintf( ...
        'powerlawTL_IC_s1_K_%s.dat', ...
        K_file_strings{i});

    data = readmatrix(filename);

    ell = data(:,2);
    Tstar = data(:,4);

    valid = isfinite(ell) & isfinite(Tstar) & ell > 0 & Tstar > 0;
    ellFit = ell(valid);
    TstarFit = Tstar(valid);

    % Equivalent to:
    % fit f(x) ... using 2:(log10($4))
    p = polyfit(ellFit, log10(TstarFit), 1);

    aTL(i) = p(1);
    bTL(i) = p(2);

    semilogy( ...
        ell, Tstar, '-o', ...
        'Color', colors(i,:), ...
        'LineWidth', 1.5, ...
        'MarkerSize', 6, ...
        'DisplayName', sprintf( ...
            '$\\widetilde{T}_{%s} \\approx 10^{%.3f}10^{%.3f\\ell}$', ...
            K_labels{i}, bTL(i), aTL(i)));

end

powerlawTL = mean(aTL);
stdevTL = std(aTL, 0);

xlabel('Domain Factor $\ell$', ...
    'Interpreter', 'latex', ...
    'FontSize', fontSize);

ylabel('Time Window of Peak Transient Energy $\widetilde{T}_{K}$', ...
    'Interpreter', 'latex', ...
    'FontSize', fontSize);

title('Isotropic Domain Time-Scaling Law for 2D Kuramoto-Sivashinsky', ...
    'Interpreter', 'latex', ...
    'FontSize', fontSize);

subtitle(sprintf([ ...
    '$K=[10^4,10^{4.5},\\ldots,10^6]$ and ' ...
    '$\\ell=[1.05,1.20,\\ldots,1.95]$: ' ...
    '$\\widetilde{T}_{K} \\sim 10^{%.4f\\ell \\pm %.4f}$'], ...
    powerlawTL, stdevTL), ...
    'Interpreter', 'latex', ...
    'FontSize', fontSize);

grid on;
box off;
xlim([1 2]);

set(gca, ...
    'XScale', 'linear', ...
    'YScale', 'log', ...
    'FontSize', fontSize);

legend( ...
    'Location', 'southeast', ...
    'FontSize', fontSize-4, ...
    'Interpreter', 'latex', ...
    'NumColumns', 3);

exportgraphics(gcf, 'powerlawTL.pdf', 'ContentType', 'vector');

fprintf('Time-vs-domain exponents:\n');
disp(aTL);
fprintf('Mean exponent = %.8f\n', powerlawTL);
fprintf('Sample std    = %.8f\n\n', stdevTL);


%% ========================================================================
%  5. SOLUTION BRANCHES
%     gnuplot: plotbranch.txt
%
%     Column 4 = time window
%     Column 7 = maximum L2 energy amplification
%
%     The original gnuplot file plots all 5 K values x all 7 ell values.
%     Curves with the same ell use the same color.
% ========================================================================

figure('Position', figPosition);
hold on;

legendHandles = gobjects(numel(ells),1);

for iK = 1:numel(K_values)

    for iEll = 1:numel(ells)

        ell = ells(iEll);

        filename = sprintf( ...
            'branch_IC_s1_K_%s_ell1_%.2f_ell2_%.2f.dat', ...
            K_file_strings{iK}, ell, ell);

        data = readmatrix(filename);

        T = data(:,4);
        Kstar = data(:,7);

        h = loglog( ...
            T, Kstar, '-o', ...
            'Color', colors(iEll,:), ...
            'LineWidth', 1.5, ...
            'MarkerSize', 5);

        % Only one legend entry per ell, matching the conceptual intent
        % of the gnuplot plot where color denotes ell.
        if iK == 1
            h.DisplayName = sprintf('%.2f', ell);
            legendHandles(iEll) = h;
        else
            h.HandleVisibility = 'off';
        end
    end
end

xlabel('Time Window', ...
    'Interpreter', 'latex', ...
    'FontSize', fontSize);

ylabel('Maximum $L^2$ Energy Amplification', ...
    'Interpreter', 'latex', ...
    'FontSize', fontSize);

title([ ...
    'Solution Branches for $K=[10^4,10^{4.5},\ldots,10^6]$ and ' ...
    '$\ell=[1.05,1.20,\ldots,1.95]$'], ...
    'Interpreter', 'latex', ...
    'FontSize', fontSize);

grid on;
box on;

set(gca, ...
    'XScale', 'log', ...
    'YScale', 'log', ...
    'FontSize', fontSize);

legend( ...
    legendHandles, ...
    'Location', 'northeast', ...
    'NumColumns', 5, ...
    'FontSize', fontSize, ...
    'Interpreter', 'latex');

exportgraphics(gcf, 'branch.pdf', 'ContentType', 'vector');


%% ========================================================================
%  Summary
% ========================================================================

fprintf('\n============================================================\n');
fprintf('SUMMARY OF FITTED SCALINGS\n');
fprintf('============================================================\n');

fprintf('K* vs K:    exponent = %.8f +/- %.8f\n', ...
    powerlawK, stdevK);

fprintf('K* vs ell:  log10 slope = %.8f +/- %.8f\n', ...
    powerlawL, stdevL);

fprintf('T* vs K:    exponent = %.8f +/- %.8f\n', ...
    powerlawTK, stdevTK);

fprintf('T* vs ell:  log10 slope = %.8f +/- %.8f\n', ...
    powerlawTL, stdevTL);

fprintf('============================================================\n');

File: Data/postprocess/generatePlots.py
#!/usr/bin/env python3
"""Generate scaling-law and solution-branch plots from 2D-KS .dat files.

Python replacement for ``generatePlots.m``. By default, the script reads data
and writes PDFs in the current directory. An alternative directory may be
given as the sole positional argument.
"""

from __future__ import annotations

import argparse
from pathlib import Path

import matplotlib

matplotlib.use("Agg")
import matplotlib.pyplot as plt
import numpy as np


ELLS = np.array([1.05, 1.20, 1.35, 1.50, 1.65, 1.80, 1.95])
K_VALUES = np.array([1.0e4, 3.2e4, 1.0e5, 3.2e5, 1.0e6])
K_FILE_STRINGS = ["1.0e+04", "3.2e+04", "1.0e+05", "3.2e+05", "1.0e+06"]
K_LABELS = [r"10^{4.0}", r"10^{4.5}", r"10^{5.0}", r"10^{5.5}", r"10^{6.0}"]
FONT_SIZE = 20
COLORS = np.array(
    [
        [0.0000, 0.4470, 0.7410],
        [0.8500, 0.3250, 0.0980],
        [0.9290, 0.6940, 0.1250],
        [0.4940, 0.1840, 0.5560],
        [0.4660, 0.6740, 0.1880],
        [0.3010, 0.7450, 0.9330],
        [0.6350, 0.0780, 0.1840],
    ]
)


def configure_matplotlib() -> None:
    plt.rcParams.update(
        {
            "font.size": FONT_SIZE,
            "axes.titlesize": FONT_SIZE,
            "axes.labelsize": FONT_SIZE,
            "legend.fontsize": FONT_SIZE - 4,
            "xtick.labelsize": FONT_SIZE,
            "ytick.labelsize": FONT_SIZE,
            "mathtext.fontset": "cm",
            "font.family": "serif",
            "pdf.fonttype": 42,
        }
    )


def load_data(path: Path) -> np.ndarray:
    data = np.genfromtxt(path, comments="#", ndmin=2)
    if data.size == 0 or data.shape[1] < 7:
        raise ValueError(f"{path} must contain at least seven numeric columns")
    return data


def fit_loglog(x: np.ndarray, y: np.ndarray) -> tuple[float, float]:
    valid = np.isfinite(x) & np.isfinite(y) & (x > 0) & (y > 0)
    if np.count_nonzero(valid) < 2:
        raise ValueError("at least two positive finite points are required for a fit")
    slope, intercept = np.polyfit(np.log10(x[valid]), np.log10(y[valid]), 1)
    return float(slope), float(intercept)


def fit_semilog_y(x: np.ndarray, y: np.ndarray) -> tuple[float, float]:
    valid = np.isfinite(x) & np.isfinite(y) & (x > 0) & (y > 0)
    if np.count_nonzero(valid) < 2:
        raise ValueError("at least two positive finite points are required for a fit")
    slope, intercept = np.polyfit(x[valid], np.log10(y[valid]), 1)
    return float(slope), float(intercept)


def new_figure() -> tuple[plt.Figure, plt.Axes]:
    return plt.subplots(figsize=(12, 8))


def finish_figure(
    fig: plt.Figure,
    ax: plt.Axes,
    output: Path,
    *,
    xlabel: str,
    ylabel: str,
    title: str,
    subtitle: str,
    legend_location: str,
    legend_columns: int = 1,
    xlim: tuple[float, float] | None = None,
    box: bool = True,
) -> None:
    ax.set_xlabel(xlabel)
    ax.set_ylabel(ylabel)
    ax.set_title(f"{title}\n{subtitle}", pad=14)
    ax.grid(True, which="major", alpha=0.35)
    ax.grid(False, which="minor")
    if not box:
        ax.spines["top"].set_visible(False)
        ax.spines["right"].set_visible(False)
    if xlim is not None:
        ax.set_xlim(*xlim)
    if legend_location == "outside below":
        ax.legend(loc="upper center", bbox_to_anchor=(0.5, -0.18), ncols=legend_columns)
    else:
        ax.legend(loc=legend_location, ncols=legend_columns)
    fig.tight_layout()
    fig.savefig(output, format="pdf", bbox_inches="tight")
    plt.close(fig)


def print_fit_summary(name: str, slopes: np.ndarray) -> tuple[float, float]:
    mean = float(np.mean(slopes))
    sample_std = float(np.std(slopes, ddof=1))
    print(f"{name}:")
    print(slopes)
    print(f"Mean exponent = {mean:.8f}")
    print(f"Sample std    = {sample_std:.8f}\n")
    return mean, sample_std


def plot_energy_vs_k(data_dir: Path, compensated: bool = False) -> tuple[float, float]:
    fig, ax = new_figure()
    slopes = np.zeros(ELLS.size)
    intercepts = np.zeros(ELLS.size)

    for i, ell in enumerate(ELLS):
        data = load_data(data_dir / f"powerlawK_IC_s1_ell1_{ell:.2f}_ell2_{ell:.2f}.dat")
        k, kstar = data[:, 0], data[:, 6]
        plotted = kstar / k**2 if compensated else kstar
        slopes[i], intercepts[i] = fit_loglog(k, plotted)
        if compensated:
            label = rf"$\widetilde K_{{{ell:.2f}}}/K^2 \approx 10^{{{intercepts[i]:.2f}}}K^{{{slopes[i]:.2f}}}$"
        else:
            label = rf"$\widetilde K_{{{ell:.2f}}} \approx 10^{{{intercepts[i]:.2f}}}K^{{{slopes[i]:.2f}}}$"
        ax.loglog(k, plotted, "-o", color=COLORS[i], linewidth=1.5, markersize=6, label=label)

    mean, std = print_fit_summary("Energy scaling K exponents", slopes)
    if compensated:
        ylabel = r"Compensated $\widetilde{K}_{\ell}/K^2$"
        subtitle = rf"$K=[10^4,\ldots,10^6]$, $\ell=[1.05,\ldots,1.95]$: $\widetilde{{K}}_\ell/K^2\sim K^{{{mean:.2f}\,\pm\,{std:.2f}}}$"
        output = data_dir / "powerlawK_comp.pdf"
        location, columns = "outside below", 3
    else:
        ylabel = r"Peak Transient Energy $\widetilde{K}_{\ell}$"
        subtitle = rf"$K=[10^4,10^{{4.5}},\ldots,10^6]$, $\ell=[1.05,1.20,\ldots,1.95]$: $\widetilde{{K}}_\ell\sim K^{{{mean:.2f}\,\pm\,{std:.2f}}}$"
        output = data_dir / "powerlawK.pdf"
        location, columns = "lower right", 1
    finish_figure(
        fig,
        ax,
        output,
        xlabel=r"Initial Energy $K$",
        ylabel=ylabel,
        title=r"$L^2$ Energy-Scaling Law for 2D Kuramoto-Sivashinsky",
        subtitle=subtitle,
        legend_location=location,
        legend_columns=columns,
        xlim=(1e4, 1e6),
    )
    return mean, std


def plot_energy_vs_ell(
    data_dir: Path, *, power_fit: bool = False, compensated: bool = False
) -> tuple[float, float]:
    fig, ax = new_figure()
    slopes = np.zeros(K_VALUES.size)
    intercepts = np.zeros(K_VALUES.size)

    for i, (k_string, k_label) in enumerate(zip(K_FILE_STRINGS, K_LABELS)):
        data = load_data(data_dir / f"powerlawL_IC_s1_K_{k_string}.dat")
        ell_squared, kstar = data[:, 1] ** 2, data[:, 6]
        plotted = kstar / ell_squared if compensated else kstar
        if power_fit:
            slopes[i], intercepts[i] = fit_loglog(ell_squared, plotted)
            prefix = r"/\ell^2" if compensated else ""
            label = rf"$\widetilde K_{{{k_label}}}{prefix}\approx 10^{{{intercepts[i]:.2f}}}\ell^{{2({slopes[i]:.2f})}}$"
        else:
            slopes[i], intercepts[i] = fit_semilog_y(ell_squared, plotted)
            label = rf"$\widetilde K_{{{k_label}}}\approx 10^{{{intercepts[i]:.2f}}}10^{{{slopes[i]:.2f}\ell^2}}$"
        ax.semilogy(ell_squared, plotted, "-o", color=COLORS[i], linewidth=1.5, markersize=6, label=label)

    mean, std = print_fit_summary("Domain scaling energy exponents", slopes)
    if compensated:
        output = data_dir / "powerlawL_comp.pdf"
        ylabel = r"Compensated $\widetilde{K}_{K}/\ell^2$"
        law = rf"$\widetilde{{K}}_K/\ell^2\sim\ell^{{2({mean:.2f}\,\pm\,{std:.2f})}}$"
    elif power_fit:
        output = data_dir / "powerlawL.pdf"
        ylabel = r"Peak Transient Energy $\widetilde{K}_{K}$"
        law = rf"$\widetilde{{K}}_K\sim\ell^{{2({mean:.2f}\,\pm\,{std:.2f})}}$"
    else:
        output = data_dir / "powerlawL_lin.pdf"
        ylabel = r"Peak Transient Energy $\widetilde{K}_{K}$"
        law = rf"$\widetilde{{K}}_K\sim 10^{{{mean:.2f}\ell^2\,\pm\,{std:.2f}}}$"
    if power_fit:
        ax.set_xscale("log")
    finish_figure(
        fig,
        ax,
        output,
        xlabel=r"Domain Factor $\ell^2$",
        ylabel=ylabel,
        title="Isotropic Domain-Scaling Law for 2D Kuramoto-Sivashinsky",
        subtitle=rf"$K=[10^4,\ldots,10^6]$, $\ell=[1.05,\ldots,1.95]$: {law}",
        legend_location="outside below",
        legend_columns=3,
        box=False,
    )
    return mean, std


def plot_universal_energy(data_dir: Path) -> tuple[float, float]:
    fig, ax = new_figure()
    slopes = np.zeros(ELLS.size)
    intercepts = np.zeros(ELLS.size)
    for i, ell in enumerate(ELLS):
        data = load_data(data_dir / f"powerlawK_IC_s1_ell1_{ell:.2f}_ell2_{ell:.2f}.dat")
        k, kstar = data[:, 0], data[:, 6]
        plotted = kstar / ell**2
        slopes[i], intercepts[i] = fit_loglog(k, plotted)
        label = rf"$\widetilde K_{{{ell:.2f}}}/\ell^2\approx 10^{{{intercepts[i]:.2f}}}K^{{{slopes[i]:.2f}}}$"
        ax.loglog(k, plotted, "-o", color=COLORS[i], linewidth=1.5, markersize=6, label=label)

    mean, std = print_fit_summary("Universal energy scaling K exponents", slopes)
    mean_intercept = float(np.mean(intercepts))
    finish_figure(
        fig,
        ax,
        data_dir / "powerlaw_universal.pdf",
        xlabel=r"Initial Energy $K$",
        ylabel=r"Normalized Energy $\widetilde{K}_{\ell}/\ell^2$",
        title="Normalized Energy-Scaling Law for 2D Kuramoto-Sivashinsky",
        subtitle=rf"$K=[10^4,\ldots,10^6]$, $\ell=[1.05,\ldots,1.95]$: $\widetilde{{K}}_\ell/\ell^2\approx 10^{{{mean_intercept:.2f}}}K^{{{mean:.2f}}}$",
        legend_location="outside below",
        legend_columns=3,
        xlim=(1e4, 1e6),
    )
    return mean, std


def plot_time_vs_k(data_dir: Path) -> tuple[float, float]:
    fig, ax = new_figure()
    slopes = np.zeros(ELLS.size)
    intercepts = np.zeros(ELLS.size)
    for i, ell in enumerate(ELLS):
        data = load_data(data_dir / f"powerlawTK_IC_s1_ell1_{ell:.2f}_ell2_{ell:.2f}.dat")
        k, tstar = data[:, 0], data[:, 3]
        slopes[i], intercepts[i] = fit_loglog(k, tstar)
        label = rf"$\widetilde T_{{{ell:.2f}}}\approx 10^{{{intercepts[i]:.2f}}}K^{{{slopes[i]:.2f}}}$"
        ax.loglog(k, tstar, "-o", color=COLORS[i], linewidth=1.5, markersize=6, label=label)
    mean, std = print_fit_summary("Time-vs-K exponents", slopes)
    finish_figure(
        fig,
        ax,
        data_dir / "powerlawTK.pdf",
        xlabel=r"Initial Energy $K$",
        ylabel=r"Time Window of Peak Transient Energy $\widetilde{T}_{\ell}$",
        title=r"$L^2$ Energy Time-Scaling Law for 2D Kuramoto-Sivashinsky",
        subtitle=rf"$K=[10^4,10^{{4.5}},\ldots,10^6]$, $\ell=[1.05,1.20,\ldots,1.95]$: $\widetilde{{T}}_\ell\sim K^{{{mean:.2f}\,\pm\,{std:.2f}}}$",
        legend_location="lower left",
        xlim=(1e4, 1e6),
    )
    return mean, std


def plot_time_vs_ell(data_dir: Path, power_fit: bool = False) -> tuple[float, float]:
    fig, ax = new_figure()
    slopes = np.zeros(K_VALUES.size)
    intercepts = np.zeros(K_VALUES.size)
    for i, (k_string, k_label) in enumerate(zip(K_FILE_STRINGS, K_LABELS)):
        data = load_data(data_dir / f"powerlawTL_IC_s1_K_{k_string}.dat")
        ell_squared, tstar = data[:, 1] ** 2, data[:, 3]
        if power_fit:
            slopes[i], intercepts[i] = fit_loglog(ell_squared, tstar)
            label = rf"$\widetilde T_{{{k_label}}}\approx 10^{{{intercepts[i]:.2f}}}\ell^{{2({slopes[i]:.2f})}}$"
        else:
            slopes[i], intercepts[i] = fit_semilog_y(ell_squared, tstar)
            label = rf"$\widetilde T_{{{k_label}}}\approx 10^{{{intercepts[i]:.2f}}}10^{{{slopes[i]:.2f}\ell^2}}$"
        ax.semilogy(ell_squared, tstar, "-o", color=COLORS[i], linewidth=1.5, markersize=6, label=label)
    mean, std = print_fit_summary("Time-vs-domain exponents", slopes)
    if power_fit:
        ax.set_xscale("log")
        output = data_dir / "powerlawTL.pdf"
        law = rf"$\widetilde{{T}}_K\sim\ell^{{2({mean:.2f}\,\pm\,{std:.2f})}}$"
    else:
        output = data_dir / "powerlawTL_lin.pdf"
        law = rf"$\widetilde{{T}}_K\sim10^{{{mean:.2f}\ell^2\,\pm\,{std:.2f}}}$"
    finish_figure(
        fig,
        ax,
        output,
        xlabel=r"Domain Factor $\ell^2$",
        ylabel=r"Time Window of Peak Transient Energy $\widetilde{T}_{K}$",
        title="Isotropic Domain Time-Scaling Law for 2D Kuramoto-Sivashinsky",
        subtitle=rf"$K=[10^4,10^{{4.5}},\ldots,10^6]$, $\ell=[1.05,1.20,\ldots,1.95]$: {law}",
        legend_location="outside below",
        legend_columns=3,
        box=False,
    )
    return mean, std


def plot_branches(data_dir: Path) -> None:
    fig, ax = new_figure()
    for i_k, k_string in enumerate(K_FILE_STRINGS):
        for i_ell, ell in enumerate(ELLS):
            data = load_data(data_dir / f"branch_IC_s1_K_{k_string}_ell1_{ell:.2f}_ell2_{ell:.2f}.dat")
            label = f"{ell:.2f}" if i_k == 0 else "_nolegend_"
            ax.loglog(data[:, 3], data[:, 6], "-o", color=COLORS[i_ell], linewidth=1.5, markersize=5, label=label)
    finish_figure(
        fig,
        ax,
        data_dir / "branch.pdf",
        xlabel="Time Window",
        ylabel=r"Maximum $L^2$ Energy Amplification",
        title=r"Solution Branches for $K=[10^4,10^{4.5},\ldots,10^6]$ and $\ell=[1.05,1.20,\ldots,1.95]$",
        subtitle="",
        legend_location="upper right",
        legend_columns=5,
    )


def main() -> None:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("data_dir", nargs="?", type=Path, default=Path.cwd(), help="directory containing the .dat files (default: current directory)")
    args = parser.parse_args()
    data_dir = args.data_dir.expanduser().resolve()
    if not data_dir.is_dir():
        parser.error(f"not a directory: {data_dir}")

    configure_matplotlib()
    k_mean, k_std = plot_energy_vs_k(data_dir)
    plot_energy_vs_k(data_dir, compensated=True)
    ell_mean, ell_std = plot_energy_vs_ell(data_dir)
    plot_energy_vs_ell(data_dir, power_fit=True)
    plot_energy_vs_ell(data_dir, power_fit=True, compensated=True)
    plot_universal_energy(data_dir)
    tk_mean, tk_std = plot_time_vs_k(data_dir)
    tl_mean, tl_std = plot_time_vs_ell(data_dir)
    plot_time_vs_ell(data_dir, power_fit=True)
    plot_branches(data_dir)

    print("\n============================================================")
    print("SUMMARY OF FITTED SCALINGS")
    print("============================================================")
    print(f"K* vs K:    exponent = {k_mean:.8f} +/- {k_std:.8f}")
    print(f"K* vs ell:  log10 slope = {ell_mean:.8f} +/- {ell_std:.8f}")
    print(f"T* vs K:    exponent = {tk_mean:.8f} +/- {tk_std:.8f}")
    print(f"T* vs ell:  log10 slope = {tl_mean:.8f} +/- {tl_std:.8f}")
    print("============================================================")


if __name__ == "__main__":
    main()
